### ISSUE TITLE
Occlusion query node doesn't work well far from the origin

### ISSUE THREAD

**Issue by gedalia** [openscenegraph/OpenSceneGraph#538]
Had to make a number of changes to allow it to work far from the origin, otherwise z value transformations started to give poor results and yield false negatives.

https://github.com/openscenegraph/OpenSceneGraph/blob/ae24be56f30261aae7f9f86039ec9eac96d4c3c9/src/osg/OcclusionQueryNode.cpp#L573

   Vec3 base = (*v)[0];
           for (int i = 0; i < 8; i++) {
              (*v)[i] -= base;
           }
        
           osg::Matrix matrix;
           matrix.makeTranslate(base);
           nonConstThis->setBBoxTransform(matrix);

           Geometry* query_geom = static_cast<Geometry*>(nonConstThis->_queryGeode->getDrawable(0));
           query_geom->setCullingActive(false);
           query_geom->setVertexArray(v.get());

           Geometry* debug_geom = static_cast<osg::Geometry*>(nonConstThis->_debugGeode->getDrawable(0));
           debug_geom->setCullingActive(false);
           debug_geom->setVertexArray(v.get());
           
           // use the initially calculated values.
           BoundingSphere bs;
           bs.set(cbv.getBoundingBox().center(), cbv.getBoundingBox().radius());
           return bs;

and  changes to draw implementation
 osg::Matrix old = renderInfo.getState()->getModelViewMatrix();
    osg::Matrix local_offset =  _queryNode.getBBoxTransform() * old;
    renderInfo.getState()->applyModelViewMatrix(local_offset);
    renderInfo.getState()->applyModelViewAndProjectionUniformsIfRequired();

    ext->glBeginQuery( GL_SAMPLES_PASSED_ARB, tr->_id );
    osg::Geometry::drawImplementation( renderInfo );
    ext->glEndQuery( GL_SAMPLES_PASSED_ARB );

    renderInfo.getState()->applyModelViewMatrix( old );
    renderInfo.getState()->applyModelViewAndProjectionUniformsIfRequired();

**Comment by openscenegraph** [comment]
I am afraid using snippets like this to report a problem isn't helpful for us to reproduce the issue and properly review the changes that may provide a possible fix.

The best way to discuss these issues as they arise is via the osg-users mailing list/forum, as plenty of other developers will view the discussion and be able to chip in, very few developers frequent the Issues on github in comparison.

To help others reproduce the issue it's really useful to have a small test case that can illustrate the problem, for instance modifying an existing OSG example to reproduce the problem is a great way for others to be able to jump in and test the problem and then investigate.

Finally, code snippets are hard to follow unless the code is just one or two lines changed, it's far better to just provide how files changed so we can do a direct diff, or provide a link to a commit against your own OSG github clone or via a PR to this main OSG repo if the PR is at the point of being well tested enough to be merged with master.

As things stand I can't do much further as taking a guess at what changes you've actually made to what code isn't reliable enough and I don't have a test case to work from so wouldn't know if the problem is fixed or not.

**Comment by gedalia** [comment]
It's a pretty typical precision issue, Far from the origin the points of the bounding box loose precision. Typically this is mitigated by most things drawing relative to a parent node. So most of the precision is retained in the matrix double multiplies on the cpu. IE a form of camera relative rendering. WIth the occlusion nodes the raw locations are being baked into a vbo, causing visible shuddering of the bounding box if you turn off debug line drawing and switch to filled, (I saw this in a database at 250,000, 2,500,000) The database was stable but the occlusion nodes visibly vibrated as you moved the camera. This is a pretty common result of sending a number like 2,500,000.25 to be transformed on the gpu 
This becomes much worse for an occlusion query because precision is so essential to the comparisons and occlusion decisions. 
  My Previous submissions usually got nixed so I just went with here's the conceptual problem, there's probably a right OSG way to fix it.  I wound up subtracting the minimum point from the VBO and putting that translation in matrix transform. It fixes the problem, but is a bit tricky since OSG is then trying to cull out the bounding box.. so there's probably something I missed . 
   I'm still struggling to come up with a fix to the fact that when occlusion nodes into view they get culled for a couple of frames.

**Comment by mp3butcher** [comment]
As Robert says, a simple example to reproduce the issue and dedicated topic on the forum/ML would be welcome to diagnose...
My thoughts about what I believe to understand of what you says :
-If you think it's a precision issue, why don't you try to scale your scene?
-your far plane is very far, in this case perhaps your scene is to big and you should split your scene into several PagedLod..
(But i admit I didn't really get your problem)

**Comment by gedalia** [comment]
-If you think it's a precision issue, why don't you try to scale your scene?
I know it's a precision issue, the change to a local space bounding box fixes both the vibrations (which is a high value precision issue I've seen for years. and false negatives with the occlusion test. Switching to reverse Z also helps, I'm not looking to fix my scene, I'm trying to let you guys know about a deficiency, I have it working with my hacky changes. it's the same kind of precision issue detailed here: http://forum.openscenegraph.org/viewtopic.php?t=11754&view=next

**Comment by gedalia** [comment]
This modification should break the occlusion test.. The code I included will stabilize the bounding box. Although other wierdness starts happening with this example that I'm unsure about.
OpenSceneGraph\examples\osgocclusionquery\osgocclusionquery.cpp

osg::ref_ptr<osg::Node>
createStockScene()
{
   // Create a simple box occluder
   osg::ref_ptr<osg::Group> root = new osg::Group();

   osg::ref_ptr<osg::MatrixTransform> transformNode = new osg::MatrixTransform();
   osg::Matrix translation = osg::Matrix::translate(2000000, 1, 100);
   transformNode->setMatrix(translation);
   transformNode->addChild( createBox().get() );

   root->addChild(transformNode);

   // Create a complex mess of triangles as a child below an
    //   OcclusionQueryNode. The OQN will ensure that the
    //   subgraph isn't rendered when it's not visible.
    osg::ref_ptr<osg::OcclusionQueryNode> oqn = new osg::OcclusionQueryNode;
    oqn.get()->setDebugDisplay(true);
    
    transformNode = new osg::MatrixTransform();
    transformNode->setMatrix(translation);
    transformNode->addChild(createRandomTriangles(200000).get());
    oqn->addChild(transformNode);

    root->addChild(oqn);

    return root.get();
}

**Comment by mp3butcher** [comment]
Do you have a problem understanding written English?
Explain your problem on the forum/ML with an attached cpp file!!!!
And again, it really sound like a precision issue you'll have to overcome yourself and not an osg bug

**Comment by openscenegraph** [comment]
I'm going to close this issue as despite repeated clear requests for how to progress this matter forward they have been ignored.  I really don't know what more I can do.

Please realize that code snippets are really unreliable and often very hard to piece together.  If it was clear we have repeated that the snippets aren't clear in this instance.  The patch to "fix" this issue is not clear at all.  The final code snippet for recreation the issue will need further work to graft it into other code, but you have to guess what goes where, and if we get it wrong then we have no way of knowing what we are testing is exactly what you are testing.  

FYI, osg::OcclusionsQueryNode and the associated example were written by a OSG contribution many years ago, and who has since retired completely from software, so those of us maintaining the OSG has to read the code and work out what is going, we can't just jump in a second guess things, we aren't omnipotent.  If we ask for details in a certain form it's because it's by far the best way for us to resolve the issue.

If you think this is genuine issue you'd like to see an improvement on then please raise the topic of precision and OcclussionQueryNode on the osg-users ML/forum as this is feature request.
